pennylane/decomposition/decomposition_rule.py

```diff
@@ -0,0 +1,337 @@
+        additional information relevant to its resource estimate into a compressed data structure.
+        To check what (if any) additional information is required to declare an operator type
+        in a resource function, refer to the ``resource_keys`` attribute of the operator class.
+        Operators with non-empty ``resource_keys`` must be declared using ``qml.resource_rep``,
+        with keyword arguments matching its ``resource_keys`` exactly.
+
+        .. seealso::
+
+            :func:`~pennylane.resource_rep`
+
+    """
+
+    def _decorator(_qfunc) -> DecompositionRule:
+        return DecompositionRule(_qfunc, resources)
+
+    return _decorator(qfunc) if qfunc else _decorator
+
+
+class DecompositionRule:  # pylint: disable=too-few-public-methods
+    """Represents a decomposition rule for an operator."""
+
+    def __init__(self, func: Callable, resources: Callable | dict):
+        self._impl = func
+        self._source = inspect.getsource(func)
+        if isinstance(resources, dict):
+
+            def resource_fn():
+                return resources
+
+            self._compute_resources = resource_fn
+        else:
+            self._compute_resources = resources
+
+    def __call__(self, *args, **kwargs):
+        return self._impl(*args, **kwargs)
+
+    def __str__(self):
+        return dedent(self._source).strip()
+
+    def compute_resources(self, *args, **kwargs) -> Resources:
+        """Computes the resources required to implement this decomposition rule."""
+        if self._compute_resources is None:
+            raise NotImplementedError("No resource estimation found for this decomposition rule.")
+        gate_counts = self._compute_resources(*args, **kwargs)
+        assert isinstance(gate_counts, dict), "Resource function must return a dictionary."
+        gate_counts = {_auto_wrap(op): count for op, count in gate_counts.items() if count > 0}
+        return Resources(gate_counts)
+
+
+def _auto_wrap(op_type):
+    """Conveniently wrap an operator type in a resource representation."""
+    if isinstance(op_type, CompressedResourceOp):
+        return op_type
+    if not issubclass(op_type, Operator):
+        raise TypeError(
+            "The keys of the dictionary returned by the resource function must be a subclass of "
+            "Operator or a CompressedResourceOp constructed with qml.resource_rep"
+        )
+    try:
+        return resource_rep(op_type)
+    except TypeError as e:
+        raise TypeError(
+            f"Operator {op_type.__name__} has non-empty resource_keys. A resource "
+            f"representation must be explicitly constructed using qml.resource_rep"
+        ) from e
+
+
+_decompositions = defaultdict(list)
+"""dict[type, list[DecompositionRule]]: A dictionary mapping operator types to decomposition rules."""
+
+
+def add_decomps(op: Type[Operator], *decomps: DecompositionRule) -> None:
```

Comment:
Hey , why was this reverted from `op_type`? Saying `op` when we mean a type seems rather problematically unclear.

Reply:
Actually, not as much unclear as clearly pointing to the wrong type of object.

Reply:
@isaacdevlugt seems to like `op` instead of `op_type`. I personally prefer `op_type` to be honest.

Reply:
I rather strongly prefer `op_type`, but I'm willing to be overwritten. We just keep track if this starts causing confusion and incorrect usage.

Reply:
Is `op_type` used elsewhere? 🤔

Reply:
I don't think it is, but I also haven't found another function that exclusively accepts an operator type as argument.

Reply:
But we do use `op` consistently in our code base to indicate an operator instance.  We don't really have any others that act on an operator type.

Reply:
Changed back to `op_type`